src/Pangraph.hs

```diff
@@ -1,116 +1,141 @@
 
--- import Data.List
-import qualified Data.ByteString                  as BS
-import qualified Algebra.Graph.HigherKinded.Class as H
+    -- Operations on Edge and Vertex
+    lookupVertexValues, lookupEdgeValues,
+    vertexContainsKey,  edgeContainsKey,
 
-data Pangraph = Pangraph [Node] [Edge] deriving (Eq)
-data Node = Node [Attribute] deriving (Eq)
-data Edge = Edge [Attribute] deriving (Eq)
-data Attribute = Attribute (BS.ByteString, BS.ByteString) deriving (Eq)
+    -- Utility Operations
+    vertexAssocList, edgeAssocList,
 
+) where
+
+import Data.Maybe            (catMaybes, mapMaybe, fromMaybe, isJust)
+import Data.ByteString.Char8 (pack, unpack)
+import Data.Map.Strict       (Map)
+import qualified Data.Map.Strict  as Map
+import qualified Data.ByteString  as BS
+
+data Pangraph = Pangraph
+  { vertices' :: Map VertexID Vertex
+  , edges' :: Map EdgeID Edge
+  , nextEdge' :: EdgeID
+  } deriving (Eq)
+data Vertex = Vertex
+  { vertexID' :: VertexID
+  , vertexAttributes' :: [Attribute]
+  } deriving (Eq)
+data Edge = Edge
+  { edgeID' :: Maybe EdgeID
+  , edgeAttributes' :: [Attribute]
+  , endpoints' :: (Vertex, Vertex)
+  } deriving (Eq)
+
+type EdgeID = Int
+type VertexID = BS.ByteString
+type Attribute = (Key, Value)
 type Key = BS.ByteString
 type Value = BS.ByteString
--- Pangraph [Node [Attribute ("id","0")],Node [Attribute ("id","1")]] [Edge [Attribute ("source","0"),Attribute ("target","1")]]
-
---  Instances of show
 
-instance Show (Pangraph) where
-  show p = "Pangraph " ++  (show $ nodes p) ++ " " ++ (show $ edges p)
+type MalformedEdge = (Edge, (Maybe Vertex, Maybe Vertex))
 
-instance Show (Node) where
-  -- show (Node []) = ""
-  show n = "Node " ++ (show $ attributes n) -- ++ ", " ++ (show $ tail ns)
+instance Show Pangraph where
+  show p = "makePangraph " ++ show (Map.elems (vertices' p)) ++ " " ++ show (Map.elems (edges' p))
 
-instance Show (Edge) where
-  -- show (Edge []) = ""
-  show e = "Edge " ++ (show $ attributes e) -- ++ ", " ++ (show ns) ++ "]"
+instance Show Vertex where
+  show (Vertex i as) = unwords ["makeVertex", show i, show as]
 
-instance Show Attribute where
-  show a = "Attribute " ++ (show $ pair a)
+instance Show Edge where
+  show (Edge i as e) = unwords ["makeEdge", show as, show e]
 
--- | Pangraph type getters
+-- List based constructors
 
-class HasAttribute a where
-  attributes :: a -> [Attribute]
+makePangraph :: [Vertex] -> [Edge] -> Maybe Pangraph
+makePangraph vs es = case verifyGraph vertexMap es of
+  [] -> Just $ Pangraph vertexMap edgeMap (1 + Map.size edgeMap)
+  abberrations -> Nothing
+  where
+    vertexMap :: Map VertexID Vertex
+    vertexMap = Map.fromList $ zip (map vertexID vs) vs
+    edgeMap :: Map EdgeID Edge
+    edgeMap = Map.fromList indexEdges
+    indexEdges :: [(EdgeID, Edge)]
+    indexEdges = map (\ (i, Edge _ as a) -> (i, Edge (Just i) as a )) $ zip [0..] es
+
+verifyGraph :: Map VertexID Vertex -> [Edge] -> [MalformedEdge]
```

Comment:
I'm not convinced finding lists of malformed edges is important functionality for Pangraph. Why not simply return `True` or `False`? Perhaps, give this function a more precise name too, e.g. `containsMalformedEdge`. Users can find the malformed edges themselves if need be, so this is unnecessary for the API.